src/writer.rs

```diff
@@ -0,0 +1,450 @@
+
+        Ok(())
+    }
+
+    /// Build root and leaf directories from entries.
+    /// Leaf directories are written to output.
+    /// The root directory is returned.
+    fn build_directories(&mut self) -> PmtResult<Directory> {
+        let (root_dir, num_leaves) = self.optimize_directories(MAX_INITIAL_BYTES - HEADER_SIZE)?;
+        if num_leaves > 0 {
+            // Write leaf directories
+            for leaf in root_dir.entries() {
+                let len = leaf.length as usize;
+                let mut dir = Directory::with_capacity(len);
+                for entry in self.entries.drain(0..len) {
+                    dir.push(entry);
+                }
+                dir.write_compressed_to(&mut self.out, self.header.internal_compression)?;
+            }
+        }
+        Ok(root_dir)
+    }
+
+    fn optimize_directories(&self, target_root_len: usize) -> PmtResult<(Directory, usize)> {
+        // Same logic as go-pmtiles optimizeDirectories
+        if self.entries.len() < 16384 {
+            let root_dir = Directory::from_entries(&self.entries);
+            let root_bytes = root_dir.compressed_size(self.header.internal_compression)?;
+            // Case1: the entire directory fits into the target len
+            if root_bytes <= target_root_len {
+                return Ok((root_dir, 0));
+            }
+        }
+
+        // TODO: case 2: mixed tile entries/directory entries in root
+
+        // case 3: root directory is leaf pointers only
+        // use an iterative method, increasing the size of the leaf directory until the root fits
+
+        let mut leaf_size = (self.entries.len() / 3500).max(4096);
+        loop {
+            let (root_dir, num_leaves) = self.build_roots_leaves(leaf_size)?;
+            let root_bytes = root_dir.compressed_size(self.header.internal_compression)?;
+            if root_bytes <= target_root_len {
+                return Ok((root_dir, num_leaves));
+            }
+            leaf_size += leaf_size / 5; // go-pmtiles: leaf_size *= 1.2
+        }
+    }
+    fn build_roots_leaves(&self, leaf_size: usize) -> PmtResult<(Directory, usize)> {
+        let mut root_dir = Directory::with_capacity(self.entries.len() / leaf_size);
+        let mut offset = 0;
+        for chunk in self.entries.chunks(leaf_size) {
+            let leaf_size = self.dir_size(chunk)?;
+            root_dir.push(DirEntry {
+                tile_id: chunk[0].tile_id,
+                offset,
+                length: into_u32(leaf_size)?,
+                run_length: 0,
+            });
+            offset += leaf_size as u64;
+        }
+
+        let num_leaves = root_dir.entries().len();
+        Ok((root_dir, num_leaves))
+    }
+    fn dir_size(&self, entries: &[DirEntry]) -> PmtResult<usize> {
+        let dir = Directory::from_entries(entries);
+        dir.compressed_size(self.header.internal_compression)
+    }
+
+    /// Finish writing the `PMTiles` file.
+    pub fn finish(mut self) -> PmtResult<()> {
```

Comment:
If finalizing a file is required, should we add the `Drop` support, possibly panicing on error? Otherwise we could be left in a bad non-finalized state... unless of course user ends writing early due to some other error - in which case we are also left in a bad state, but we shouldn't finalize it...   Or perhaps we should delete the output?

We could also introduce `poison` concept - if any `write` method on this struct fail, we set a flag before returning an error, and in that case the file will not be finalized automatically. Otherwise, finalize the file on Drop unless it was already finalized by explicit call.

How does std::fs  handle errors during Drop?

Reply:
I'm not sure about that. I see the point, but don't think that we should put too much effort into fixing an obvious user bug.

Reply:
ok, lets keep as is.  I tend to use the word `finalize`, but either works